### ISSUE TITLE
Unable to parse Custom ActionType

### ISSUE THREAD

**Issue by tnev** [Microsoft/AdaptiveCards#2220]
# Platform

* Any

# Author or host

host

# Version of SDK

1.1

# Issue

The library currently has a **Custom** ActionType, but the ActionParserRegistration doesn't support parsing a Custom action. If the ActionParserRegistration were to add a custom action parser to its cardElementParsers map, such as:

```swift
m_cardElementParsers.insert({{ActionTypeToString(ActionType::OpenUrl), std::make_shared<OpenUrlActionParser>()},
                                     {ActionTypeToString(ActionType::ShowCard), std::make_shared<ShowCardActionParser>()},
                                     {ActionTypeToString(ActionType::Submit), std::make_shared<SubmitActionParser>()},
                                     {ActionTypeToString(ActionType::Custom), std::make_shared<CustomActionParser>()}});
```

then it seems that it would be able to support custom actions. Is there a plan for supporting Custom actions in this way, or is there another suggested way of accomplishing this?

**Comment by RebeccaAnne** [comment]
You can add a custom action parser using the ActionParserRegistration::AddParser method, passing the name of your custom action and something that implements ActionElementParser. This will add your custom parser to that array.

**Comment by tnev** [comment]
Hi @RebeccaAnne , thanks for your response. This is actually something I was trying to do before, but for some reason it doesn't seem like ActionParserRegistration is publicly exposed through the iOS framework (sorry, I'm realizing now I should have specified iOS as the platform above!).

I thought it was possibly an issue with the way my project imported the framework, but I wasn't able to access it in the iOS example project you provide either. When I tried to import "AdaptiveCards/ActionParserRegistration.h" I received compiler errors about some of the frameworks included in the pch.h file under the AdaptiveSharedNamespace definition.

Its seems like the Android framework exposes it so I'm not sure if this is an issue with how I'm trying to access it or if there are limitations to accessing the C++ files on iOS. I appreciate your help!

**Comment by nesalang** [comment]
Hi @tnev, iOS doesn't have custom parser supports for actions. 
we plan to add support for custom action parsing.
It will be supported in the similar way as in base cards elements. 
When unknown type is encountered during parsing, adaptivecards create an Unknown element that stores the raw json enclosed in the custom type.
ACRCustomRenderer is a proxy that get called when the unknown type  element is encountered during rendering. From here, you can parse the raw json and render UI elements. To support more than one type, you can branch out to different types from here. 

One limitation is that you won't be able to access other parsers, so you can't make custom collection type object that needs to have access to BaseCard element parsing. E.g., custom columnset that has column and textblocks in the column.

Please let me know if this would work for you and your team.

Thanks,

**Comment by shawnphoffman** [comment]
@nesalang - Do you have an idea of when you'll be releasing full support for custom actions? @tnev and I are part of a team that is working with Adaptive Cards internally and we planned on extending them to help make them a little more powerful for our specific use case. For example, given the following card: 

```
{
  "$schema": "http://adaptivecards.io/schemas/adaptive-card.json",
  "type": "AdaptiveCard",
  "version": "1.1",
  "body": [
    {
      "type": "Container",
      "items": [
        {
          "type": "TextBlock",
          "text": "Sample TextBlock",
          "size": "medium"
        }
      ]
    }
  ],
  "actions": [
    {
      "type": "Action.Alert",
      "title": "Alert Action",
      "data": {
        "foo": "bar",
        "val": 123,
        "bool": true
      }
    }
  ]
}
```

With the adaptive cards JS library, we can register a custom action handler and pretty much implement whatever we need it to do: 

```
AdaptiveCards.AdaptiveCard.actionTypeRegistry.registerType('Action.Alert', () => {
	return new CustomAlertAction()
})
```

CustomAlertAction extends AdaptiveCards.Action allowing it to parse and validate incoming action data among other things. We're also able to handle the execution of the action to handle it in whatever way we deem fit. In the case of CustomAlertAction, we can display a browser alert with data from the action and card itself.

*Will anything like this be possible in the near future for iOS?* Support seems to exist for the Android SDK although not as straightforward as the JS implementation but we haven't been able to sort out any way to make it work for iOS.

**Comment by nesalang** [comment]
This scenario will be supported but since it's near the end of the year, it will likely happen after December.

**Comment by nesalang** [comment]
@tnev & @shawnphoffman Thanks for waiting. Could you please try this branch (jwoo/ios-custom-element-update)? 
#2299 
You can find the usage by following `CustomActionNewType`

**Comment by nesalang** [comment]
fixed by #2325
